cmd/ephemeral-installer-register-cluster-and-infraenv/main.go

```diff
@@ -0,0 +1,244 @@
+	}
+	pullSecret := secret.StringData[".dockerconfigjson"]
+
+	releaseImageVersion, releaseImageCPUArch := getReleaseVersionAndCpuArch(getReleaseVersion())
+	log.Info("releaseImage: " + getReleaseVersion())
+	log.Infof("releaseImage version %s cpuarch %s", releaseImageVersion, releaseImageCPUArch)
+
+	var infraEnv aiv1beta1.InfraEnv
+	if infraenvErr := getFileData(Options.InfraEnvFile, &infraEnv); infraenvErr != nil {
+		log.Fatal(infraenvErr.Error())
+	}
+
+	var nmStateConfig aiv1beta1.NMStateConfig
+	if nmStateErr := getFileData(Options.NMStateConfigFile, &nmStateConfig); nmStateErr != nil {
+		log.Fatal(nmStateErr.Error())
+	}
+
+	staticNetworkConfig, processErr := processNMStateConfig(log, infraEnv, nmStateConfig)
+	if processErr != nil {
+		log.Fatal(processErr.Error())
+	}
+
+	clientConfig := client.Config{}
+	u, parseErr := url.Parse(Options.ServiceBaseUrl)
+	if parseErr != nil {
+		log.Fatal(parseErr, "Failed parsing inventory URL")
+	}
+	u.Path = path.Join(u.Path, client.DefaultBasePath)
+	clientConfig.URL = u
+	bmInventory := client.New(clientConfig)
+	ctx := context.Background()
+
+	log.Info("SERVICE_BASE_URL: " + Options.ServiceBaseUrl)
+	log.Info("Registering cluster")
+
+	clusterParams := controllers.CreateClusterParams(&cd, &aci, pullSecret, releaseImageVersion, releaseImageCPUArch, nil)
+	clientClusterParams := &installer.V2RegisterClusterParams{
+		NewClusterParams: clusterParams,
+	}
+	clusterResult, registerClusterErr := bmInventory.Installer.V2RegisterCluster(ctx, clientClusterParams)
+	if registerClusterErr != nil {
+		log.Fatal(registerClusterErr, "Failed to register cluster with assisted-service")
+	}
+
+	modelsCluster := clusterResult.Payload
+
+	log.Info("Registered cluster with id: " + modelsCluster.ID.String())
+
+	log.Info("Registering infraenv")
+
+	infraEnvParams := controllers.CreateInfraEnvParams(&infraEnv, models.ImageType(Options.ImageTypeISO), pullSecret, modelsCluster.ID, modelsCluster.OpenshiftVersion)
+
+	if len(staticNetworkConfig) > 0 {
+		log.Infof("Added %d nmstateconfigs", len(staticNetworkConfig))
+		infraEnvParams.InfraenvCreateParams.StaticNetworkConfig = staticNetworkConfig
+	}
+
+	clientInfraEnvParams := &installer.RegisterInfraEnvParams{
+		InfraenvCreateParams: infraEnvParams.InfraenvCreateParams,
+	}
+	infraEnvResult, registerInfraEnvErr := bmInventory.Installer.RegisterInfraEnv(ctx, clientInfraEnvParams)
+	if registerInfraEnvErr != nil {
+		log.Fatal(registerInfraEnvErr, "Failed to register infraenv with assisted-service")
+	}
+
+	modelsInfraEnv := infraEnvResult.Payload
+
+	log.Info("Registered infraenv with id: " + modelsInfraEnv.ID.String())
+}
+
+func getReleaseVersion() string {
+	var clusterImageSet hivev1.ClusterImageSet
+	if err := getFileData(Options.ClusterImageSetFile, &clusterImageSet); err != nil {
+		log.Fatal(err.Error())
+	}
+	return clusterImageSet.Spec.ReleaseImage
+}
+
+func getReleaseVersionAndCpuArch(releaseImage string) (version string, cpuArch string) {
```

Comment:
Would be nice to have explicit failure when you're given a release image that doesn't contain the architecture in the tag - or alternatively, default to x86

Reply:
We will need to validate the release image URL at a higher level before it reaches this client. Like you said, by the time it reaches this client it will be too late. I'm not sure if defaulting to x86_64 is a safe assumption. Let me ask the team.

Reply:
We avoid parsing the ocp release image uri by extracting the version/arch from the image.
See: https://github.com/openshift/assisted-service/blob/f74ba9bdd99924146140c448982c3b925c2669d3/internal/versions/versions.go#L297-L307
This allows us to support a release image uri with a commit hash, i.e. we return the ocp version and cpu arch defined in the image metadata.

Reply:
Good point about commit hash. I'll switch to using the release package to determine the image version and architecture.